### PR TITLE
Upgrade mongodb

### PR DESCRIPTION
**Pull request by eeerin**

### CODE REVIEW

File: lib/db.js [hunk 1]
```diff
@@ -36,7 +37,40 @@ const actionOneOrMany = curry(async (action, collection, isMultiple, ...params)
...
+    return new Date(data.valueOf());
+  }
+
+  const keys = Object.keys(data);
+  if (keys.length === 1) {
+    if (keys[0] === '$oid') {
+      return new ObjectId(data[keys[0]]);
+    }
+
+    if (keys[0] === '$date') {
```
Comment (spacetc62) [inline lib/db.js:65]: do we actually need to do $date? I'd almost rather fix the places where we are handing in $date

File: lib/db.js [hunk 2]
```diff
@@ -36,7 +37,40 @@ const actionOneOrMany = curry(async (action, collection, isMultiple, ...params)
...
+const deserialize = (data = {}) => {
+  if (Array.isArray(data)) {
+    return data.map(deserialize);
+  }
+  if (data === null) { return; }
+  if (typeof data !== 'object' || data === null) {
+    return data;
+  }
+
+  if (data.__type === 'bson.ObjectID') {
```
[the remark this answers is not in the document]
Comment (spacetc62) [reply lib/db.js:51]: or, just straight always unwrap immutables in general (i.e., call .toObject())

File: lib/model.js [hunk 3]
```diff
@@ -291,14 +288,14 @@ const _create = (data, definition) => {
...
-    return _saveHelper(this, options, null);
+    return _saveHelper(this, options);
   };
 
   definition.methods.saveWithId = async function(id, options) {
     if (!this.isNew()) {
       throw new Error('saveWithId must receive a newly created object');
     }
-    return _saveHelper(this, options, new ObjectID(id));
+    return _saveHelper(this, options, id);
```
Comment (spacetc62) [inline lib/model.js:298]: why this change?

File: test/model.js [hunk 4]
```diff
@@ -234,13 +232,13 @@ describe('Model', () => {
 
   describe('.saveWithId()', function() {
     it('should save with a given ID', async () => {
-      const id = new bson.ObjectID('1234abcd103f8e485c9d2019');
+      const id = new bson.ObjectID('1234abcd103f8e485c9d2017');
```
Comment (spacetc62) [inline test/model.js:235]: why this change?

File: test/model.js [hunk 5]
```diff
@@ -423,7 +421,7 @@ describe('Model', () => {
       const count = await FullTestModel.count({ _id: updatedModel._id });
       expect(count).to.equal(1);
       const collection = _db.collection('mocha_test');
-      const doc = await collection.findOne({ _id: updatedModel._id });
+      const doc = await collection.findOne({ _id: updatedModel._id.toObject() });
```
Comment (spacetc62) [inline test/model.js:424]: seems like with your deserialize, you don't need this unwrap?

File: test/model.js [hunk 6]
```diff
@@ -439,7 +437,7 @@ describe('Model', () => {
       expect(updatedModel._id.toString()).to.equal(newModel._id.toString());
       expect(updatedModel._etag).to.not.equal(newModel._etag);
       const collection = _db.collection('mocha_test');
-      const doc = await collection.findOne({ _id: updatedModel._id });
+      const doc = await collection.findOne({ _id: updatedModel._id.toObject() });
```
Comment (spacetc62) [inline test/model.js:440]: same here?

**Review by spacetc62 — CHANGES_REQUESTED** [review]
there is an EJSON serialize in model save helper - I think that might not make sense to have anymore?

File: lib/cursor.js [hunk 7]
```diff
@@ -82,17 +82,17 @@ class Cursor {
   /**
    * Returns `true` if the cursor is closed.
    * @returns {boolean} `true` if the cursor is closed.
-   * @see {@link http://mongodb.github.io/node-mongodb-native/3.0/api/Cursor.html#isClosed}
+   * @see {@link https://mongodb.github.io/node-mongodb-native/4.0/classes/findcursor.html#close}
    */
   isClosed() {
-    return this._cursor.isClosed();
+    return this._cursor.close;
```
[the remark this answers is not in the document]
Comment (eeerin) [reply lib/cursor.js:88]: oops yes you are right. I noticed this change in the platform with the cursor-iterator helpers in losant-common

File: lib/db.js [hunk 8]
```diff
@@ -203,14 +236,25 @@ class Database extends EventEmitter {
...
     debug('getting db collection');
-    docs = EJSON.deserialize(docs);
+    docs = deserialize(docs);
     const collection = await this.collection(collectionName);
-    return insertOneOrMany(collection, Array.isArray(docs), docs);
+    const results = await insertOneOrMany(collection, Array.isArray(docs), docs);
+
+    if (Array.isArray(docs)) {
+      docs.forEach((doc, i) => {
+        doc._id = results.insertedIds[`${i}`];
```
[the remark this answers is not in the document]
Comment (eeerin) [reply lib/db.js:250]: if I'm reading this code correctly yes https://github.com/mongodb/node-mongodb-native/blob/b67af3cd/src/operations/insert.ts#L29 

The index is a string not a number as the key on the object to the inserted object id

File: test/model.js (hunk 4 above)
Comment (eeerin) [reply test/model.js:235]: I changed this when I was getting duplicate key errors I'll revert.

File: lib/db.js [hunk 9]
```diff
@@ -36,7 +37,39 @@ const actionOneOrMany = curry(async (action, collection, isMultiple, ...params)
...
 const insertOneOrMany = actionOneOrMany('insert');
 const updateOneOrMany = actionOneOrMany('update');
-const removeOneOrMany = actionOneOrMany('remove');
+const deleteOneOrMany = actionOneOrMany('delete');
+
+const deserialize = (data = {}) => {
+  if (Array.isArray(data)) {
+    return data.map(deserialize);
+  }
+  if (data === null) { return; }
```
Comment (spacetc62) [inline lib/db.js:46]: this should return null

File: lib/db.js [hunk 10]
```diff
@@ -36,7 +37,39 @@ const actionOneOrMany = curry(async (action, collection, isMultiple, ...params)
...
+  if (typeof data !== 'object' || data instanceof Date) {
```
Comment (spacetc62) [inline lib/db.js:47]: instead of this, I'd to:

```
if(typeof data !== 'object'){
  return data;
}
if (Immutable.isImmutable(data)) {
  return data.toObject();
}
if(!isPlainObject(data)){
  return data;
}
```

And that will catch Date, as well as things like RegExp or instances of ObjectID or other weird objects we don't actually want to iterate

File: lib/model.js [hunk 11]
```diff
@@ -5,19 +5,17 @@ const debug = require('debug')('cosa:model');
...
-
 const { buildPropertySchema } = require('./utils');
-
+const { ObjectId } = require('bson');
 const Immutable = require('./immutable');
 Immutable.use(require('./defined-object'));
 Immutable.use('array', require('./array'));
 Immutable.use(Date, require('./date'));
-Immutable.use('objectid', require('./bson-objectId'));
+Immutable.use('Objectid', require('./bson-objectId'));
```
Comment (spacetc62) [inline lib/model.js:18]: why did this change? and why did ObjectID change to ObjectId?

**Review by spacetc62 — CHANGES_REQUESTED** [review]

File: lib/db.js [hunk 12]
```diff
@@ -1,12 +1,13 @@
 const debug = require('debug')('cosa:db');
 const defaults = require('defaults');
-const EJSON = require('mongodb-extended-json');
+const { ObjectId } = require('bson');
```
Comment (spacetc62) [inline lib/db.js:3]: is there a difference between requiring ObjectID vs ObjectId? is one the correct one vs the other?

**Review by spacetc62 — CHANGES_REQUESTED** [review]

File: lib/db.js (hunk 12 above)
Comment (eeerin) [reply lib/db.js:3]: Nope exactly the same. ObjectId gets exported as both ObjectId and ObjectID. 
https://github.com/mongodb/js-bson/blob/master/src/bson.ts#L84

Comment (spacetc62) [reply lib/db.js:3]: looking it looks like they would prefer you use `ObjectId` - is that correct? if so we should just go ahead and make everything in here consistent with ObjectId

Comment (eeerin) [reply lib/db.js:3]:
```
// In 4.0.0 and 4.0.1, this property name was changed to ObjectId to match the class name.
  // This caused interoperability problems with previous versions of the library, so in
  // later builds we changed it back to ObjectID (capital D) to match legacy implementations.
 ```
 
 kind of seems like it doesnt matter to them. but I guess we can to keep it consistent.

**Review by spacetc62 — APPROVED** [review]
